app-misc/anki: add 23.12.1

antecrescent opened the pull request:
I took great inspiration from [NixOS' Anki package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/anki/default.nix). I'd be happy about some feedback!
Should I copy the `gl-fixup` patch from NixOS? I've had no problems with the GL library so far.
I'll gladly co-maintain this package if need be.

TODO:
- [x] ~`bzip2_sys` appears to compile the library statically, why?~
The crate was dropped with version 2.1.66.
- [x] ~Is `dev-libs/openssl` needed?~
The OpenSSL library is linked against _rsbridge.so only when testing, otherwise it uses Rustls.

thesamesam commented on app-misc/anki/metadata.xml:
```diff
@@ -13,4 +13,7 @@
 	<upstream>
```
please co-maintain the package too, thanks!

antecrescent replied:
Sure!

antecrescent commented on app-misc/anki/anki-2.1.66.ebuild:
```diff
@@ -0,0 +1,675 @@
+LICENSE="AGPL-3+ Apache-2.0 BSD CC-BY-3.0 GPL-3+ MIT"
+
+SLOT="0"
+KEYWORDS="~amd64"
```
Yes, unfortunately, I don't have a dedicated testing environment yet and I don't own an x86 machine.

parona-source replied:
To my understanding you still keep keywords on upgrades even if you can't test it. You would only drop them if you had to add dependencies which weren't keyworded. 

https://devmanual.gentoo.org/keywording/index.html#keywording-on-upgrades

antecrescent replied:
Readded keyword `~x86` in 1dd818ebd0d7664794141c32591d21d9c567df63

thesamesam commented on app-misc/anki/anki-2.1.66.ebuild:
```diff
@@ -0,0 +1,675 @@
+pkg_setup() {
+	python-single-r1_pkg_setup
+	# The Python and Yarn environment variables prevent a download of the binaries
+	export PYTHON_BINARY=/bin/false YARN_BINARY=/bin/false \
```
It's okay to rely on PATH lookups for things like `false`, but not so much for things in `BDEPEND`.

antecrescent replied:
Implemented in fa3d721b6df32178926fe4b12abc1b91b21f460c

antecrescent commented on app-misc/anki/anki-2.1.66.ebuild:
```diff
@@ -0,0 +1,675 @@
+# PyQt6: Actually need bindings for QtGui with PyQt_OpenGL enabled,
+# that is qtbase[opengl,gui] and PyQt6[gui], not PyQt6[gui,opengl].
+# dev-qt/qtgui:5 enables opengl support by default.
+# If !qt6 block PyQt6, because of the Qt6 detection mechanism
```
I think this should do it: fa3d721b6df32178926fe4b12abc1b91b21f460c

antecrescent commented on app-misc/anki/anki-2.1.66.ebuild:
```diff
@@ -0,0 +1,675 @@
+	eapply_user
+}
+
+src_compile() {
```
Implemented in fa3d721b6df32178926fe4b12abc1b91b21f460c. Also exported `AR`, `CC`, `PKG_CONFIG` (abd0f6b0ac6784e8732bc557bc75563088924bef) like cargo eclass as some crates compile non-Rust code etc.

antecrescent commented:
ping. I implemented the requested changes last week. Is there anything else I can do apart from squashing and rebasing?

parona-source commented:
Oh this has gone for far too long. 

@thesamesam ?

Squashh + rebase would be good though. Better history for gentoo.

antecrescent commented:
I squashed and rebased the commit.
Changes to last iteration:
* complete LICENSE statement
* add missing dependencies
    - checkboxes in settings require `PyQt{5,6}[svg]` at runtime
    - `qt/tools/build_ui.py`  requires `PyQt{5,6}` at build time
* install MIME types file into `/usr/share/mime/packages/`

I tested the package against py3.12 but `qtwebengine:6` doesn't support it yet.

antecrescent commented:
Changes to last iteration:
* add missing bug tags to commit message

parona-source commented:
> I tested the package against py3.12 but qtwebengine:6 doesn't support it yet.

You only have to care about if PyQt6WebEngine supports it. The python compat in qtwebengine:6 is about using python in the build system, it doesn't provide python bindings itself.

PythonCompatUpdate: version 2.1.66: PYTHON_COMPAT update available: python3_12

parona-source commented on app-misc/anki/anki-2.1.66.ebuild:
```diff
@@ -0,0 +1,724 @@
+# dev-qt/qtgui:5 enables opengl support by default.
+
+DEPEND="
+	dev-db/sqlite
+	>=app-arch/zstd-1.5.2
```
```suggestion
	dev-db/sqlite:3
	>=app-arch/zstd-1.5.2:=
```

```
 * VDB: detected possibly incorrect RDEPEND (app-misc/anki-2.1.66)
 * app-arch/zstd | app-arch/zstd:=
 * dev-db/sqlite | dev-db/sqlite:3
```

```
ldd /usr/lib/python3.11/site-packages/anki/_rsbridge.so
        linux-vdso.so.1 (0x00007fff97131000)
        libzstd.so.1 => /lib64/libzstd.so.1 (0x00007fc15bd40000)
        libsqlite3.so.0 => /usr/lib64/libsqlite3.so.0 (0x00007fc15a600000)
        libgcc_s.so.1 => /usr/lib/gcc/x86_64-pc-linux-gnu/13/libgcc_s.so.1 (0x00007fc15bd16000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fc15a920000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fc15a417000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc15be68000)
        libicui18n.so.73 => /usr/lib64/libicui18n.so.73 (0x00007fc15a000000)
        libicuuc.so.73 => /usr/lib64/libicuuc.so.73 (0x00007fc159c00000)
        libstdc++.so.6 => /usr/lib/gcc/x86_64-pc-linux-gnu/13/libstdc++.so.6 (0x00007fc159800000)
        libicudata.so.73 => /usr/lib64/libicudata.so.73 (0x00007fc157800000)
```

parona-source commented:
```
ask@bigglane /home/ask
 $ quse -DvIp anki
app-misc/anki-2.1.66
  lto*                              <no description found>
  qt6*                              Add support for the Qt 6 application and UI framework
  test                              Enable dependencies and/or preparations necessary to run tests (usually controlled by FEATURES=test but can be toggled independently)
  debug                             Enable extra debug codepaths, like asserts and extra output. If you want to get meaningful backtraces see https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Backtraces
  python_single_target_python3_10   Build for Python 3.10 only
  python_single_target_python3_11*  Build for Python 3.11 only
ask@bigglane /home/ask
 $ anki
Running with temporary Qt5 compatibility shims.
Run with DISABLE_QT5_COMPAT=1 to confirm compatibility with Qt6.
Preparing to run...
Qt warning: QGuiApplication::setDesktopFileName: the specified desktop file name ends with .desktop. For compatibility reasons, the .desktop suffix will be removed. Please specify a desktop file name without .desktop suffix 
Qt debug: doh set to ""  --  SystemOnly 
Starting main loop...
```

Despite enabling qt6 it still falls back on qt5 by default.

Could either set the compat variable in env or patch it.

parona-source commented:
Using python3.12 I noticed this `orjson is missing; DB operations will be slower`. This is due to orjson missing the python3.12 target. But this would make a good optfeature addition nevertheless.

antecrescent commented:
> ```
> ask@bigglane /home/ask
>  $ anki
> Running with temporary Qt5 compatibility shims.
> Run with DISABLE_QT5_COMPAT=1 to confirm compatibility with Qt6.
> Preparing to run...
> Qt warning: QGuiApplication::setDesktopFileName: the specified desktop file name ends with .desktop. For compatibility reasons, the .desktop suffix will be removed. Please specify a desktop file name without .desktop suffix 
> Qt debug: doh set to ""  --  SystemOnly 
> Starting main loop...
> ```

With `qt6` enabled: 
```python
try:
    import PyQt6
except:
    from .qt5 import *  # type: ignore
else:
    if not os.getenv("DISABLE_QT5_COMPAT"):
        print("Running with temporary Qt5 compatibility shims.")
        print("Run with DISABLE_QT5_COMPAT=1 to confirm compatibility with Qt6.")
        from . import qt5_compat  # needs to be imported first
    from .qt6 import *
```
> 
> Despite enabling qt6 it still falls back on qt5 by default.

If I understand correctly, that is not what's happening here. At runtime, only the `try` and `else` branches are executed. The environment variable enables or disables a compatibility layer for add-ons which have not yet migrated to Qt6. [`qt5_compat.py`](https://github.com/ankitects/anki/blob/2.1.66/qt/aqt/qt/qt5_compat.py) does not import any Qt5 modules but provides aliases to their Qt6 counterparts and stubs removed functionality.
See also:
https://archive.ph/Nw9c7#95% (2.1.50 change log describing the shims)
https://archive.ph/hWzNX (Anki forum thread about the shims)
> Could either set the [compat variable in env](https://devmanual.gentoo.org/tasks-reference/environment/index.html) or patch it.

I suggest leaving it as is but I'm open to discussions :)

parona-source commented:
Ah, I misunderstood the message. Its fine then.

antecrescent commented:
Changelog to last iteration:
* add suggested slot operators `dev-db/sqlite:3`, `>=app-arch/zstd-1.5.2:=`
* test and enable py3.12 (add backport fix of runtime crash due to removed `distutils` module)
* add optional runtime dependency `dev-python/orjson` for faster database operations
* update test dependency `app-text/texlive[extra]` to `app-text/texlive`

Question:
> Using python3.12 I noticed this `orjson is missing; DB operations will be slower`. This is due to orjson missing the python3.12 target. But this would make a good optfeature addition nevertheless.

I assume enabling the py3.12 target for anki was the correct way to go as `orjson` is optional. Is this correct?

parona-source commented:
> I assume enabling the py3.12 target for anki was the correct way to go as orjson is optional. Is this correct?

Yes, and orjson got python3.12 enabled a couple hours later. So its double correct.

parona-source commented:
@thesamesam Any outstanding issues?

antecrescent commented:
Changes to last iteration:
* 2.1.66 -> 23.10.1 (update and new versioning scheme)
* Add new USE flag `minimal` which only builds Anki's sync server without the GUI.
* Update `LICENSE` to include bundled runtime JS library licenses
* Update dependencies
  * Add missing `!qt6? ( dev-qt/qtgui[jpeg,png] )` dependency
  * Relax `dev-util/cargo-nextest` and tighten `virtual/rust` dependencies
* Add documentation

Anki 23.10.1 adds the ability to change the video driver at runtime.
I added the dependencies for the Vulkan driver as an `optfeature` since I assume most users won't tinker with this option.
The dependencies do include `dev-qt/qtwebengine:6[vulkan]` though, which can take a considerable time to rebuild. Should I add the dependencies unconditionally or leave it as is?

thesamesam commented:
Thanks for the ping earlier - I'm swamped atm, but I agree this needs doing. Can you remind me in about two weeks please? Do really appreciate the patience.

antecrescent commented:
Changes to last iteration:
- Drop dependency on Python bindings for QtSvg, instead depend on QtSvg directly
- Fix typo and do some stylistic changes in `DOC_CONTENTS`

@thesamesam Your reminder, as requested! Not sure if you read my dm on IRC last weekend, so I thought I'd ping you publicly just in case :P (Thanks a lot for reviewing this rather massive ebuild.)

PIPIPIG233666 commented:
@antecrescent hey there, good work.
I didn't notice you had a source-built anki ebuild ready for review till I wrote my own based on the latest 23.12.1 release.
The patch is linked [here](https://[github.com/gentoo/gentoo/compare/master...PIPIPIG233666:gentoo:anki-23.12.1.patch](https://github.com/gentoo/gentoo/compare/master...PIPIPIG233666:gentoo:anki-23.12.1.patch))
I have several notable differences:

 - configure
	 - stop setting RELEASE for non debug/lto build, default should be 1 anyway
	 - use local binary variables instead of masking them (kanged from archlinux aur)
 - compile
	 - for minimal, only build with anki-sync-server as the target
	 - use runner instead of ninja to build
	 - build directly instead of the default wheels
- install
	 - install anki-sync-server direcly
	 - fixup wrong symlink for _rsbridge (doesn't seem to be a problem for wheels build though)
	 - only install necessary directories for python modules
 -  postinstall
	 - return 0 for minimal build

I am also thinking of adding a systemd/initrd file for the sync server.

thesamesam commented on app-misc/anki/anki-23.10.1.ebuild:
```diff
@@ -0,0 +1,801 @@
+	dev-db/sqlite:3
+	>=app-arch/zstd-1.5.5:=
+"
+GUI="
```
Let's do something like `GUI_DEPEND`. `GUI` is a bit too short and ambiguous.

thesamesam commented on app-misc/anki/anki-23.10.1.ebuild:
```diff
@@ -0,0 +1,801 @@
+RDEPEND="
+	app-misc/ca-certificates
+	!minimal? ( ${GUI} )
+	${DEPEND}
```
Put DEPEND above the contents

thesamesam commented on app-misc/anki/anki-23.10.1.ebuild:
```diff
@@ -0,0 +1,801 @@
+			  qt6? ( $(python_gen_cond_dep '>=dev-python/PyQt6-6.5.0[${PYTHON_USEDEP}]') )
+			  !qt6? ( $(python_gen_cond_dep '>=dev-python/PyQt5-5.15.5[${PYTHON_USEDEP}]') )
+			  ${PYTHON_DEPS}
+			  )
```
Fix indentation

antecrescent replied:
https://github.com/gentoo/gentoo/blob/90530f31413ac1558e36b47123186b392c451fc9/app-misc/anki/anki-23.12.1.ebuild#L791-L792
Emacs' *DEPEND indentation started to get frustrating so I switched to vim. I hope the new indentation is ok.

thesamesam commented on app-misc/anki/anki-23.10.1.ebuild:
```diff
@@ -0,0 +1,801 @@
+	>=virtual/rust-1.71.0
+	!minimal? (
+			  app-arch/unzip
+			  dev-util/ninja
```
Change to `app-alternatives/ninja`

thesamesam commented on app-misc/anki/metadata.xml:
```diff
@@ -13,4 +21,8 @@
 	</upstream>
+	<use>
+		<flag name="lto">Build with link-time optimization. Ignored if debug is enabled.</flag>
+		<flag name="minimal">Only build the standalone sync server. Does not respect the lto flag.</flag>
```
We really try to avoid USE=minimal if we can, as it's non-intuitive by name, and has confusing semantics. It's "negative USE logic".

Can't we do the positive version? USE=gui or whatever?

antecrescent commented:
Changes to last iteration:
- redo `-flto` behavior (https://github.com/gentoo/gentoo/pull/34695)
- redo patches based on https://github.com/freebsd/freebsd-ports/tree/main/games/anki
 Drop a bunch of exported environment variables
 Respect more user-controlled variables

antecrescent commented on app-misc/anki/anki-23.12.1.ebuild:
```diff
@@ -0,0 +1,955 @@
+		mkdir -p out/pyenv/bin || die
+		ln -s "${PYTHON}" out/pyenv/bin/"${EPYTHON}" || die
+		ln -s "${EPYTHON}" out/pyenv/bin/python || die
+		# TODO: ln -s "${BROOT}/usr/bin/protoc-gen-mypy" out/pyenv/bin || die
```
Atm, pseudo-venv.patch removes the code that generates stubs for Anki's protobuf files. I figured that might be ok for now, since it's "only" type hints that are missing.
I have an ebuild for mypy-protobuf on my disk but I struggle to set up the tests correctly, so I haven't published it yet.

antecrescent commented on app-misc/anki/anki-23.12.1.ebuild:
```diff
@@ -0,0 +1,955 @@
+	#    * generate the ninja file and run ninja afterwards
+	#    * create the Python wheel files in "${S}"/out/wheels
+
+	cargo_src_compile $(usev debug "--release") --package runner
```
Is is acceptable to explicitly set the `--release` flag this way? Unfortunately, rustc complains about it being set more than once.

thesamesam replied:
Do you mean:
```suggestion
	cargo_src_compile $(usev !debug "--release") --package runner
```
?

antecrescent replied:
No, I mean to run:
```suggestion
	cargo build --release --package runner || die
```

thesamesam replied:
Huh, okay.

antecrescent commented:
Thank you for the review and the suggestions!
> The patch is linked [here](https://%5Bgithub.com/gentoo/gentoo/compare/master...PIPIPIG233666:gentoo:anki-23.12.1.patch%5D(https://github.com/gentoo/gentoo/compare/master...PIPIPIG233666:gentoo:anki-23.12.1.patch))

Unfortunately, the link is broken. I assume you meant to link to your PR? 
> - configure
>   - stop setting RELEASE for non debug/lto build, default should be 1 anyway

https://github.com/ankitects/anki/blob/1a1d4d5419c6b57ef3baf99c9d2d9cf85d36ae0a/build/ninja_gen/src/build.rs#L357-L363
No, you get a debug build if `RELEASE` is not set.
>   - use local binary variables instead of masking them (kanged from archlinux aur)

Can you explain what you mean by masking, please? Maybe it doesn't apply anymore since the 23.12.1 ebuild is quite different from the 23.10.1 one.
> - compile
>   - for minimal, only build with anki-sync-server as the target

AFAIK `cargo build --manifest-path rslib/sync` = `cargo build --package anki-sync-server` but I agree that the latter variant is much less cryptic. I changed it :)
>  - use runner instead of ninja to build

Done.
>  - build directly instead of the default wheels

Why? Upstream recommends that you build and install the wheels into your local venv. Also, from what I understand, pip uses dist-info metadata from wheels to resolve dependencies (see: https://bugs.gentoo.org/920330). 
> - install
>   - install anki-sync-server direcly

If you install the GUI, the sync server is baked into the rsbridge backend. You can even call it directly, so `anki-sync-server` would be a duplicate. https://docs.ankiweb.net/sync-server.html#from-a-packaged-build
>  - fixup wrong symlink for _rsbridge (doesn't seem to be a problem for wheels build though)

I don't know yet, what's going on here.
>  - only install necessary directories for python modules

Looking at your ebuild, why did you choose not to install dist-info?
> * postinstall
>   * return 0 for minimal build

I don't understand this.

> I am also thinking of adding a systemd/initrd file for the sync server.

That's a good idea :+1:

thesamesam commented on app-misc/anki/files/23.12.1/ninja-gentoo-setup.patch:
```diff
@@ -0,0 +1,80 @@
+  Can also be implemented in the ebuild via runner args when the rest
+  of the patch gets dropped.
+* Respect NINJA_STATUS if already set.
+* Respect app-alternatives/ninja symlink (n2 is not yet in the tree).
```
FWIW, happy to consider a PR for [this](https://github.com/evmar/n2) separately.

antecrescent replied:
Once upstream tags a release, I'd take you up on this.

thesamesam commented on app-misc/anki/files/23.12.1/remove-yarn-dep.patch:
```diff
@@ -0,0 +1,68 @@
+This is an adapted patch from the games/anki FreeBSD port.
```
Could you add links to where you got it from/the inspiration? In general, the more references the better, even if you adapted it or only used it as inspiration.
